keps/sig-node/4381-dra-semantic-parameters/README.md

```diff
@@ -0,0 +1,1477 @@
+    selector: |-
+      quantities["memory"] <= "16Gi"
+```
+
+In this example, the additional selector expression limits users of this class
+to smaller sub-partitions. Together with limiting the number of claims that
+users are allowed to create for this class (see resource quotas in the core
+KEP) this can ensure that users do not consume too much resources. Allowing
+resource quotas that are based on resource attributes may be a useful future
+enhancement.
+
+### Communicating allocation to the DRA driver
+
+The scheduler decides which resources to use for a claim and how much of
+them. It also needs to pass through the opaque vendor parameters, if there are
+any. All of this information gets stored in the allocation result inside the
+ResourceClaim status. For the example above, the result produced by the
+scheduler is simply the list of IDs of the selected sub-partitions:
+
+```yaml
+# Matches with the SemanticResourceHandle Go type defined below.
+vendorClassParameters:
+  ...
+vendorClaimParameters:
+  kind: CardParameters
+  apiVersion: dra.example.com/v1alpha1
+  metadata:
+    name: my-parameters
+    namespace: user-namespace
+    uid: foobar-uid
+  ...
+  spec:
+    count: 2
+    minimumRuntimeVersion: v12.0.0
+    minimumMemory: 8Gi
+    sharing:
+      strategy: TimeSliced
+
+nodeName: worker-1
+partitionable:
+  instances:
+  - partitionable-gpu-2/half-and-quarters/second-half/quarter-0
+  - partitionable-gpu-2/half-and-quarters/second-half/quarter-1
+```
+
+## Design Details
+
+### ResourceClass extension
+
+A new, optional field in ResourceClass enables semantic parameters for claims
+using this class:
+
+```go
+type ResourceClass struct {
+    ...
+
+    // If (and only if) allocation of claims using this class is handled
+    // via semantic parameters, then SemanticParameters must be set to true.
+    SemanticParameters bool
+}
+```
+
+### NodeResourceSlice
+
+For each node, one or more NodeResourceSlice objects get created. The kubelet
+publishes them with the node as the owner, so they get deleted when a node goes
+down and then gets removed.
+
+All list types are atomic because that makes tracking the owner for
+server-side-apply (SSA) simpler. Patching individual list elements is not
+needed and there is a single owner (kubelet).
+
+```go
+// NodeResourceSlice provides information about available
+// resources on individual nodes.
+type NodeResourceSlice struct {
+    metav1.TypeMeta
+    // Standard object metadata
+    // +optional
+    metav1.ObjectMeta
+
+    Spec NodeResourceSliceSpec
+}
+```
+
+There is only a spec for a NodeResourceSlice at the moment. It holds the
+information about available resources. A status is not strictly needed because
+the information in the allocated claim statuses is sufficient to determine
+which of those resources are reserved for claims.
+
+However, despite the finalizer on the claims it could happen that a
+sufficiently determined user deletes a claim while it is in use. Therefore
```

Comment:
"sufficiently determined" :laughing:

Reply:
Hah, my joke made @johnbelamaric laugh. Mission accomplished. We all need a good laugh from time to time.

I'll replace with the more neutral phrasing suggested by @thockin.

Reply:
Oh, the phrasing is fine.  It made me chuckle, too